sdks/java/io/singlestore/src/main/java/org/apache/beam/sdk/io/singlestore/Write.java

```diff
@@ -0,0 +1,255 @@
+
+  @Override
+  public PDone expand(PCollection<T> input) {
+    DataSourceConfiguration dataSourceConfiguration =
+        Util.getRequiredArgument(
+            getDataSourceConfiguration(), "withDataSourceConfiguration() is required");
+    String table = Util.getRequiredArgument(getTable(), "withTable() is required");
+    UserDataMapper<T> userDataMapper =
+        Util.getRequiredArgument(getUserDataMapper(), "withUserDataMapper() is required");
+    int batchSize = Util.getArgumentWithDefault(getBatchSize(), DEFAULT_BATCH_SIZE);
+    checkArgument(batchSize > 0, "batchSize should be greater then 0");
+
+    input
+        .apply(
+            ParDo.of(
+                new DoFn<T, List<String>>() {
+                  @ProcessElement
+                  public void processElement(ProcessContext context) throws Exception {
+                    context.output(userDataMapper.mapRow(context.element()));
+                  }
+                }))
+        .setCoder(ListCoder.of(StringUtf8Coder.of()))
+        .apply(ParDo.of(new BatchFn<>(batchSize)))
+        .apply(ParDo.of(new WriteFn<Void>(dataSourceConfiguration, table)))
+        .setCoder(VoidCoder.of());
+
+    return PDone.in(input.getPipeline());
+  }
+
+  private static class BatchFn<ParameterT> extends DoFn<ParameterT, Iterable<ParameterT>> {
+    List<ParameterT> batch = new ArrayList<>();
+    int batchSize;
+
+    BatchFn(int batchSize) {
+      this.batchSize = batchSize;
+    }
+
+    @ProcessElement
+    public void processElement(ProcessContext context) {
+      batch.add(context.element());
+      if (batch.size() >= batchSize) {
+        context.output(batch);
+        batch = new ArrayList<>();
+      }
+    }
+
+    @FinishBundle
+    public void finish(FinishBundleContext context) {
+      if (batch.size() > 0) {
+        context.output(batch, Instant.now(), GlobalWindow.INSTANCE);
+        batch = new ArrayList<>();
+      }
+    }
+  }
+
+  private static class WriteFn<OutputT> extends DoFn<Iterable<List<String>>, OutputT> {
+    DataSourceConfiguration dataSourceConfiguration;
+    String table;
+
+    WriteFn(DataSourceConfiguration dataSourceConfiguration, String table) {
+      this.dataSourceConfiguration = dataSourceConfiguration;
+      this.table = table;
+    }
+
+    @ProcessElement
+    public void processElement(ProcessContext context) throws Exception {
+      DataSource dataSource = dataSourceConfiguration.getDataSource();
+
+      Connection conn = dataSource.getConnection();
+      try {
+        Statement stmt = conn.createStatement();
+        try (PipedOutputStream baseStream = new PipedOutputStream();
+            InputStream inputStream = new PipedInputStream(baseStream, BUFFER_SIZE)) {
+          ((com.singlestore.jdbc.Statement) ((DelegatingStatement) stmt).getInnermostDelegate())
+              .setNextLocalInfileInputStream(inputStream);
+
+          final Exception[] writeException = new Exception[1];
+
+          Thread dataWritingThread =
```

Comment:
Avoid creating threads inside of dofns. In Beam, asynchronous and multithreading behavior is best handled by the Beam framework itself.

Reply:
Hmm....
I don't know how to implement this asynchronous task using Beam framework.
Here we have two parts - statement execution and writing data to the stream. 
They are tightly coupled (writing will get stuck if the statement is not executed and the buffer is full and the statement execution won't finish until writing is finished).
And also these parts share a common PipedOutputStream which is not serializable.

Does the Beam framework have features that will allow something like this?

Reply:
This is somewhat unusual for me as well. @lukecwik do we have a pattern for something like this?